### ISSUE TITLE
Traits instead of abstract listener classes

### ISSUE THREAD

**Issue by leofeyer** [contao/core-bundle#256]
Currently we have two abstract listener classes, `AbstractScopeAwareListener` and `AbstractUserAwareListener`, whereby the user listener extends from the scope listener. Both add horizontal and not vertical functionality so they actually should be traits, don't you think?

https://github.com/contao/core-bundle/blob/feature/symfony-session/src/EventListener/AbstractUserAwareListener.php

https://github.com/contao/core-bundle/blob/feature/symfony-session/src/EventListener/AbstractScopeAwareListener.php

Symfony has added a `ContainerAwareTrait` some time ago for the exact same purpose.

**Comment by aschempp** [comment]
Probably true. We could also use the traits for non listener classes (like the UserProvider).

**Comment by leofeyer** [comment]
Fixed in 4efc1e54351ee8807ebb8b46cb4487cdda2c859b.

**Comment by aschempp** [comment]
Where did you actually add the traits?

**Comment by leofeyer** [comment]
In `src/EventListener/`.

**Comment by aschempp** [comment]
That means we *cant* use it in other places. It would make sense for the `ContaoUserProvider` too (see https://github.com/contao/core-bundle/blob/develop/src/Security/User/ContaoUserProvider.php#L92).

I would suggest to move the traits to the `DependencyInjection` namespace.

**Comment by leofeyer** [comment]
Why did you chose `DependencyInjection`?

BTW, we had this discussion when I was adding traits in the `Traits` namespace two weeks ago. Back then, @Toflar and you held the opinion that the traits should be as specific as possible and therefore you wanted me to move them into the `Event` namespace (where I was using them). Have you changed your point of view?

**Comment by aschempp** [comment]
> Why did you chose DependencyInjection?

Because it is somehow a dependency and it is always injected using the DI configuration. The `ContainerAwareTrait` is also in the `DependencyInjection` namespace (though I agree that might also be because it's related to the service container).

> Have you changed your point of view?

I have not, because these things were very much related to events only.

**Comment by leofeyer** [comment]
Hm, I'll probably regret saying this, but a `Framework` namespace would solve this problem :)

```
src/
    Framework/
        ContaoFramework.php
        FrameworkAwareTrait.php
```

**Comment by leofeyer** [comment]
I have fixed the first part in 19033be9652c46ec35373ce536834ee0fa15aa67.
